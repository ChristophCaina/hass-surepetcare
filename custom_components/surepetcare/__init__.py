from homeassistant import core


async def async_setup_entry(hass: core.HomeAssistant, entry):
    """Set up SurePetCare from a config entry."""

    hass.async_create_task(
        hass.config_entries.async_forward_entry_setup(entry, "sensor")
    )


    return True
