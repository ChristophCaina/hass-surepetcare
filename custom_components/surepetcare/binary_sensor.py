from homeassistant.helpers.entity import BinarySensorEntity


class FeederBatterySensor(BinarySensorEntity):
    """Representation of a battery sensor for a feeder."""

    def __init__(self, coordinator, feeder):
        self.coordinator = coordinator
        self.feeder = feeder

    @property
    def name(self):
        return f"Feeder {self.feeder['name']} Battery"

    @property
    def is_on(self):
        return self.coordinator.data["battery_status"] == "ok"

    async def async_update(self):
        """Fetch new data from the API."""
        await self.coordinator.async_request_refresh()
