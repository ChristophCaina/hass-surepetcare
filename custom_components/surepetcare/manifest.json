{
  "codeowners": ["FredrikM97"],
  "config_flow": true,
  "integration_type": "hub",
  "dependencies": [],
  "documentation": "https://github.com/FredrikM97/hass-surepetcare",
  "issue_tracker": "https://github.com/FredrikM97/hass-surepetcare/issues",
  "domain": "surepetcare",
  "iot_class": "cloud_polling",
  "name": "SurePetCare",
  "requirements": ["py-surepetcare>=0.5.7-dev0"],
  "loggers": ["surepcio"],
  "version": "1.4.8-dev1"
}