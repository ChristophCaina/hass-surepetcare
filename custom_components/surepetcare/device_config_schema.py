from voluptuous import Schema, Required
from surepetcare.enums import ProductId
from homeassistant.data_entry_flow import section

DEVICE_CONFIG_SCHEMAS = {
    ProductId.FEEDER_CONNECT: {
        "schema": Schema({
            Required("location"): str,
        }),
        "title": "Feeder Connect Configuration",
    },
    ProductId.DUAL_SCAN_PET_DOOR: {
        "schema": section(Schema({
            Required("location_inside"): str,
            Required("location_outside"): str,
        })),
        "title": "Dual Scan Pet Door Configuration",
    },
}

