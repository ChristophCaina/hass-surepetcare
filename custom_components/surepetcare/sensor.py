from homeassistant.components.sensor import (
    SensorEntity,
)
from homeassistant.helpers.entity_platform import AddEntitiesCallback
from surepetcare.enums import ProductId

class SurePetcareBaseSensor(SensorEntity):
    """Base class for SurePetcare sensors with device_info."""

    def __init__(self, coordinator, device, data):
        self.coordinator = coordinator
        self.device = device

    @property
    def device_info(self):
        return {
            "identifiers": {("surepetcare", self.device["id"])},
            "name": self.device.get("name"),
            "manufacturer": "Sure Petcare",
            "model": self.device.get("product_id"),
        }

class PetLocationSensor(SensorEntity):
    """Representation of a pet's location sensor."""

    def __init__(self, coordinator, flow, data):
        self.coordinator = coordinator
        self.flow = flow
        self._data = data

    @property
    def name(self):
        return f"sensor.{self.data['id']}_location"

    @property
    def state(self):
        return self.coordinator.data["location"]

    @property
    def extra_state_attributes(self):
        # Todo: Map to data values
        return {
            "location_inside": self.flow.get("location_inside"),
            "location_outside": self.flow.get("location_outside"),
        }
    
    @property
    def unit_of_measurement(self):
        return "Location"

    async def async_update(self):
        """Fetch new data from the API."""
        await self.coordinator.async_request_refresh()


class PetLastFedSensor(SensorEntity):
    """Representation of a pet's last fed sensor."""

    def __init__(self, coordinator, flow, data):
        self.coordinator = coordinator
        self.flow = flow
        self._data = data

    @property
    def name(self):
        return f"sensor.{self.data['id']}_last_fed"

    @property
    def state(self):
        return self.coordinator.data["last_fed"]

    @property
    def unit_of_measurement(self):
        return "Time"

    async def async_update(self):
        """Fetch new data from the API."""
        await self.coordinator.async_request_refresh()


class BatterySensor(SensorEntity):
    """Representation of a Device's battery sensor."""

    def __init__(self, coordinator, device, data):
        self.coordinator = coordinator
        self.device = device

    @property
    def name(self):
        return f"sensor.{self.data['id']}_battery"

    @property
    def state(self):
        return self.coordinator.data.get("battery")

    @property
    def unit_of_measurement(self):
        return "%"

    async def async_update(self):
        """Fetch new data from the API."""
        await self.coordinator.async_request_refresh()


# Map product IDs to their sensors
PRODUCT_SENSOR_MAPPINGS = {
    ProductId.FEEDER_CONNECT: {
        "last_fed": PetLastFedSensor,
    },
    ProductId.DUAL_SCAN_PET_DOOR: {
        "location": PetLocationSensor,
        "battery": BatterySensor,
    },
}

 # Main entry: devices is a list
async def async_setup_entry(hass, entry, async_add_entities: AddEntitiesCallback):
    coordinator = hass.data.get("surepetcare_coordinator")
    entities = []
    
    for subentry in entry.subentries.values():
        product_id = subentry.data['product_id']
        for device_id, device_data in subentry.data['device'].items():
            sensor_map = PRODUCT_SENSOR_MAPPINGS.get(product_id, {})
            data = find_device_by_id(entry.data['devices'], device_id)
            for sensor_cls in sensor_map.values():
                entities.append(sensor_cls(coordinator, device_data, data))

    async_add_entities(entities)

def find_device_by_id(devices, device_id):
    return next((d for d in devices if str(d['id']) == device_id), None)
