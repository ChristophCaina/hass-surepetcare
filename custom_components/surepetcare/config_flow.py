"""Config flow for SurePetCare integration."""

from copy import deepcopy
import logging
from typing import Any
from enum import IntEnum

from surepcio import SurePetcareClient
from surepcio import Household

from surepcio.enums import ProductId
import voluptuous as vol

from homeassistant import config_entries
from homeassistant.helpers.device_registry import callback
from homeassistant.helpers.selector import selector

from .const import (
    CONF_EMAIL,
    CONF_PASSWORD,
    DOMAIN,
    POLLING_SPEED,
    SCAN_INTERVAL,
)
from .device_config_schema import DEVICE_CONFIG_SCHEMAS

logger = logging.getLogger(__name__)

STEP_USER_DATA_SCHEMA = vol.Schema(
    {
        vol.Required(CONF_EMAIL): str,
        vol.Required(CONF_PASSWORD): str,
    }
)


class FlowAction(IntEnum):
    """Determine if setting up or reconfiguring."""

    SETUP = 0
    RECONFIGURE = 1


class SurePetCareConfigFlow(config_entries.ConfigFlow, domain=DOMAIN):  # type: ignore
    """Handle a config flow for SurePetCare integration."""

    VERSION = 1

    def __init__(self) -> None:
        self.client: SurePetcareClient | None = None
        self._entities: dict[str, Any] = {}

    async def async_step_user(self, user_input: dict[str, Any] | None = None):
        """Handle the initial step to login the user"""
        errors = {}
        if user_input is not None:
            email = user_input.get(CONF_EMAIL)
            password = user_input.get(CONF_PASSWORD)
            (
                token,
                client_device_id,
                entity_info,
                error,
            ) = await self._async_fetch_entities(email=email, password=password)
            if error:
                errors["base"] = error
            else:
                return self.async_create_entry(
                    title="SurePetCare Devices",
                    data={
                        "token": token,
                        "client_device_id": client_device_id,
                        "entities": entity_info,
                    },
                )
        return self.async_show_form(
            step_id="user",
            data_schema=STEP_USER_DATA_SCHEMA,
            errors=errors,
        )

    async def _async_fetch_entities(
        self,
        email: str | None = None,
        password: str | None = None,
        token: str | None = None,
        device_id: str | None = None,
    ):
        """Authenticate and fetch devices/pets, return (token, client_device_id, entity_info, error)."""
        self.client = SurePetcareClient()
        logged_in = await self.client.login(
            email=email, password=password, token=token, device_id=device_id
        )
        if not logged_in:
            return None, None, None, "auth_failed"
        token = getattr(self.client, "token", None)
        if not token:
            return None, None, None, "cannot_connect"
        households = await self.client.api(Household.get_households())
        self._entities = {}
        for household in households:
            self._entities.update(
                {
                    str(device.id): device
                    for device in await self.client.api(household.get_devices())
                }
            )
            self._entities.update(
                {
                    str(device.id): device
                    for device in await self.client.api(household.get_pets())
                }
            )
        await self.client.close()
        if not self._entities:
            return None, None, None, "no_devices_or_pet_found"
        entity_info = {
            str(device.id): {
                "product_id": getattr(device, "product_id", None),
                "name": getattr(device, "name", str(device.id)),
            }
            for device in self._entities.values()
        }
        return token, self.client.device_id, entity_info, None

    async def async_step_reconfigure(self, user_input: dict[str, Any] | None = None):
        """Migration step in case entities not populated/new device added."""
        entry = self.hass.config_entries.async_get_entry(self.context["entry_id"])
        token = entry.data["token"]
        device_id = entry.data["client_device_id"]

        _, _, entity_info, _ = await self._async_fetch_entities(
            token=token, device_id=device_id
        )

        self.hass.config_entries.async_update_entry(
            entry,
            data={
                **entry.data,
                "entities": entity_info,
            },
        )

        return self.async_abort(reason="entities_reconfigured")

    @staticmethod
    @callback
    def async_get_options_flow(config_entry):
        """Return the options flow handler."""
        return SurePetCareOptionsFlow(config_entry)


class SurePetCareOptionsFlow(config_entries.OptionsFlowWithReload):
    """Options flow for SurePetCare integration."""

    _device_id: str | None

    def __init__(self, config_entry: config_entries.ConfigEntry) -> None:
        self._options = deepcopy(dict(config_entry.options))

    async def async_step_init(self, user_input: dict[str, Any] | None = None):
        """Show device selection menu."""

        if user_input is not None:
            if user_input.get("finished"):
                # End the flow and save current options
                return self.async_create_entry(data=self._options)
            self._device_id = user_input["device_option"]
            return await self.async_step_configure_device()

        entities = self.config_entry.data.get("entities", {})
        if not entities:
            return self.async_abort(reason="no_devices_or_pet_found")

        sorted_entities = sorted(
            entities.items(), key=lambda item: item[1].get("product_id", 0)
        )
        select_options = [
            {
                "value": str(k),
                "label": get_device_attr(v, "name", str(k))
                + f" ({ProductId(v['product_id']).name})",
            }
            for k, v in sorted_entities
        ]
        schema = vol.Schema(
            {
                vol.Required("device_option"): selector(
                    {"select": {"options": select_options}}
                ),
                vol.Optional("finished", default=False): bool,
            }
        )

        return self.async_show_form(
            step_id="init",
            data_schema=schema,
        )

    async def async_step_configure_device(
        self, user_input: dict[str, Any] | None = None
    ):
        """Configure the selected device."""

        if user_input is not None:
            self._options[self._device_id] = user_input
            return await self.async_step_init()

        entity = self.config_entry.data.get("entities", {})[self._device_id]
        schema_info = DEVICE_CONFIG_SCHEMAS.get(get_device_attr(entity, "product_id"))
        existing_config = self._options.get(self._device_id, {})
        schema_dict = {}

        if schema_info:
            for key, field_type in schema_info.items():
                default_value = existing_config.get(key) if existing_config else None
                if default_value is not None:
                    schema_dict[type(key)(key.schema, default=default_value)] = (
                        field_type
                    )
                else:
                    schema_dict[key] = field_type

        # Add polling speed option (in seconds), default SCAN_INTERVAL
        polling_default = existing_config.get(POLLING_SPEED, SCAN_INTERVAL)
        schema_dict[vol.Optional(POLLING_SPEED, default=polling_default)] = vol.All(
            int, vol.Range(min=5, max=86400)
        )

        schema = vol.Schema(schema_dict)

        return self.async_show_form(
            step_id="configure_device",
            data_schema=schema,
            description_placeholders={"device_name": entity["name"]},
        )


def get_device_attr(device: Any, attr: str, default: Any = None) -> Any:
    """Get attribute or dict key from device."""
    if isinstance(device, dict):
        return device.get(attr, default)
    return getattr(device, attr, default)
