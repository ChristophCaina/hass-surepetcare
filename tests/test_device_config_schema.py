import pytest
from custom_components.surepetcare import device_config_schema
from custom_components.surepetcare.const import LOCATION_INSIDE, LOCATION_OUTSIDE
from custom_components.surepetcare.device_config_schema import DEVICE_CONFIG_SCHEMAS
from surepcio.enums import ProductId
import voluptuous as vol


def test_dual_scan_pet_door_schema_valid():
    schema = DEVICE_CONFIG_SCHEMAS[ProductId.DUAL_SCAN_PET_DOOR]
    assert isinstance(schema, dict)
    valid = {LOCATION_INSIDE: "Hall", LOCATION_OUTSIDE: "Garden"}
    vol_schema = vol.Schema(schema, extra=vol.REMOVE_EXTRA)
    assert vol_schema(valid) == valid


def test_dual_scan_pet_door_schema_invalid():
    schema = DEVICE_CONFIG_SCHEMAS[ProductId.DUAL_SCAN_PET_DOOR]
    vol_schema = vol.Schema(schema, extra=vol.REMOVE_EXTRA)
    with pytest.raises(vol.Invalid):
        vol_schema({LOCATION_INSIDE: 123})
    with pytest.raises(vol.Invalid):
        vol_schema({LOCATION_OUTSIDE: 123123})


def test_dual_scan_pet_door_schema_missing_keys():
    schema = DEVICE_CONFIG_SCHEMAS[ProductId.DUAL_SCAN_PET_DOOR]
    vol_schema = vol.Schema(schema, extra=vol.REMOVE_EXTRA)
    assert vol_schema({}) == {}
    assert vol_schema({LOCATION_INSIDE: "Hall"}) == {LOCATION_INSIDE: "Hall"}
    assert vol_schema({LOCATION_OUTSIDE: "Garden"}) == {LOCATION_OUTSIDE: "Garden"}
    valid = {
        LOCATION_INSIDE: "Hall",
        LOCATION_OUTSIDE: "Garden",
        "extra": "ignored",
    }
    assert vol_schema({k: valid[k] for k in [LOCATION_INSIDE, LOCATION_OUTSIDE]}) == {
        LOCATION_INSIDE: "Hall",
        LOCATION_OUTSIDE: "Garden",
    }


def test_dual_scan_pet_door_schema_extra_keys_ignored():
    schema = DEVICE_CONFIG_SCHEMAS[ProductId.DUAL_SCAN_PET_DOOR]
    vol_schema = vol.Schema(schema, extra=vol.REMOVE_EXTRA)
    valid = {
        LOCATION_INSIDE: "Hall",
        LOCATION_OUTSIDE: "Garden",
        "extra": "ignored",
    }
    result = vol_schema(valid)
    assert "extra" not in result
    assert result == {LOCATION_INSIDE: "Hall", LOCATION_OUTSIDE: "Garden"}


def test_dual_scan_pet_door_schema_type_validation():
    schema = DEVICE_CONFIG_SCHEMAS[ProductId.DUAL_SCAN_PET_DOOR]
    vol_schema = vol.Schema(schema, extra=vol.REMOVE_EXTRA)
    # Both values must be str
    valid = {LOCATION_INSIDE: "Room", LOCATION_OUTSIDE: "Yard"}
    assert vol_schema(valid) == valid
    # Wrong type for location_inside
    with pytest.raises(vol.Invalid):
        vol_schema({LOCATION_INSIDE: 123, LOCATION_OUTSIDE: "Yard"})
    # Wrong type for location_outside
    with pytest.raises(vol.Invalid):
        vol_schema({LOCATION_INSIDE: "Room", LOCATION_OUTSIDE: 456})


def test_device_config_schemas_structure():
    # Ensure DEVICE_CONFIG_SCHEMAS exists and is a dict
    assert isinstance(device_config_schema.DEVICE_CONFIG_SCHEMAS, dict)
    # Check all expected ProductIds are present
    for pid in [
        ProductId.FEEDER_CONNECT,
        ProductId.DUAL_SCAN_PET_DOOR,
        ProductId.HUB,
        ProductId.PET,
    ]:
        assert pid in device_config_schema.DEVICE_CONFIG_SCHEMAS


def test_dual_scan_pet_door_schema_keys():
    schema = device_config_schema.DEVICE_CONFIG_SCHEMAS[ProductId.DUAL_SCAN_PET_DOOR]
    assert isinstance(schema, dict)
    assert LOCATION_INSIDE in [k.schema for k in schema.keys()]
    assert LOCATION_OUTSIDE in [k.schema for k in schema.keys()]


def test_schema_type_and_structure():
    # All schemas should be dict or None
    for pid, schema in DEVICE_CONFIG_SCHEMAS.items():
        assert schema is None or isinstance(schema, dict)
        if isinstance(schema, dict):
            for k, v in schema.items():
                # Keys should be voluptuous marker objects
                assert hasattr(k, "schema")
                # Values should be type or callable
                assert callable(v) or isinstance(v, type)


def test_schema_keys_are_unique():
    # If schema is a dict, keys should be unique
    for pid, schema in DEVICE_CONFIG_SCHEMAS.items():
        if isinstance(schema, dict):
            keys = [k.schema for k in schema.keys()]
            assert len(keys) == len(set(keys))


def test_schemas_are_none():
    for pid in [ProductId.FEEDER_CONNECT, ProductId.HUB, ProductId.PET]:
        assert device_config_schema.DEVICE_CONFIG_SCHEMAS[pid] is None
