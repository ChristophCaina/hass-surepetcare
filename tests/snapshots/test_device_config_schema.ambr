# serializer version: 1
# name: test_snapshot_schema
  dict({
    <ProductId.PET: 0>: dict({
      'polling_speed': All(<class 'int'>, Range(min=5, max=86400, min_included=True, max_included=True, msg=None), msg=None),
    }),
    <ProductId.HUB: 1>: dict({
      'polling_speed': All(<class 'int'>, Range(min=5, max=86400, min_included=True, max_included=True, msg=None), msg=None),
    }),
    <ProductId.PET_DOOR: 3>: dict({
      'location_inside': <class 'str'>,
      'location_outside': <class 'str'>,
      'polling_speed': All(<class 'int'>, Range(min=5, max=86400, min_included=True, max_included=True, msg=None), msg=None),
    }),
    <ProductId.FEEDER_CONNECT: 4>: dict({
      'polling_speed': All(<class 'int'>, Range(min=5, max=86400, min_included=True, max_included=True, msg=None), msg=None),
    }),
    <ProductId.DUAL_SCAN_CONNECT: 6>: dict({
      'location_inside': <class 'str'>,
      'location_outside': <class 'str'>,
      'polling_speed': All(<class 'int'>, Range(min=5, max=86400, min_included=True, max_included=True, msg=None), msg=None),
    }),
    <ProductId.POSEIDON_CONNECT: 8>: dict({
      'polling_speed': All(<class 'int'>, Range(min=5, max=86400, min_included=True, max_included=True, msg=None), msg=None),
    }),
    <ProductId.DUAL_SCAN_PET_DOOR: 10>: dict({
      'location_inside': <class 'str'>,
      'location_outside': <class 'str'>,
      'polling_speed': All(<class 'int'>, Range(min=5, max=86400, min_included=True, max_included=True, msg=None), msg=None),
    }),
    <ProductId.NO_ID_DOG_BOWL_CONNECT: 32>: dict({
      'polling_speed': All(<class 'int'>, Range(min=5, max=86400, min_included=True, max_included=True, msg=None), msg=None),
    }),
  })
# ---
# name: test_snapshot_schema_with_example_values
  dict({
    <ProductId.PET: 0>: dict({
      'polling_speed': 300,
    }),
    <ProductId.HUB: 1>: dict({
      'polling_speed': 300,
    }),
    <ProductId.PET_DOOR: 3>: dict({
      'polling_speed': 300,
    }),
    <ProductId.FEEDER_CONNECT: 4>: dict({
      'polling_speed': 300,
    }),
    <ProductId.DUAL_SCAN_CONNECT: 6>: dict({
      'polling_speed': 300,
    }),
    <ProductId.POSEIDON_CONNECT: 8>: dict({
      'polling_speed': 300,
    }),
    <ProductId.DUAL_SCAN_PET_DOOR: 10>: dict({
      'location_inside': 'Hall',
      'location_outside': 'Garden',
      'polling_speed': 300,
    }),
    <ProductId.NO_ID_DOG_BOWL_CONNECT: 32>: dict({
      'polling_speed': 300,
    }),
  })
# ---
# name: test_snapshot_schema_with_missing_values
  dict({
    <ProductId.PET: 0>: dict({
      'polling_speed': 300,
    }),
    <ProductId.HUB: 1>: dict({
      'polling_speed': 300,
    }),
    <ProductId.PET_DOOR: 3>: dict({
      'polling_speed': 300,
    }),
    <ProductId.FEEDER_CONNECT: 4>: dict({
      'polling_speed': 300,
    }),
    <ProductId.DUAL_SCAN_CONNECT: 6>: dict({
      'polling_speed': 300,
    }),
    <ProductId.POSEIDON_CONNECT: 8>: dict({
      'polling_speed': 300,
    }),
    <ProductId.DUAL_SCAN_PET_DOOR: 10>: dict({
      'polling_speed': 300,
    }),
    <ProductId.NO_ID_DOG_BOWL_CONNECT: 32>: dict({
      'polling_speed': 300,
    }),
  })
# ---
